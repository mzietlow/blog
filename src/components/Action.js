import React from 'react'
import PropTypes from 'prop-types'

const Action = (props) => {
  return (
    <div>
      <button onClick={props.handlePick} disabled={!props.hasOptions}>
        What should I do?
      </button>
    </div>
  );
};

Action.propTypes = {
  handlePick: PropTypes.func.isRequired,
  hasOptions: PropTypes.bool.isRequired,
};


export default Action;