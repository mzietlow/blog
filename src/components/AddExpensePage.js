import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import { PropTypes } from "prop-types";
import { useNavigate } from "react-router-dom";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.navigate("/");
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}
const AddExpensePageWrapper = (props) => {
  const navigate = useNavigate();
  return <AddExpensePage {...props} navigate={navigate}/>
};

AddExpensePage.propTypes = {
  startAddExpense: PropTypes.func.isRequired,
  navigate: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePageWrapper);
