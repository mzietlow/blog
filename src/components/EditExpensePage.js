import React from "react";
import PropTypes from "prop-types";
import ExpenseForm from "./ExpenseForm";
import { useSelector, connect } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expenseUpdate) => {
    this.props.startEditExpense(this.props.expense.id, expenseUpdate);
    this.props.navigate("/");
  };
  onRemoveClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.navigate("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button key={this.props.expense.id} onClick={this.onRemoveClick}>
          Remove
        </button>
      </div>
    );
  }
}

const EditExpensePageWrapper = (props) => {
  const navigate = useNavigate();
  const expenseId = useParams().id;
  const expense = useSelector((state) =>
    state.expenses.find((expense) => expense.id === expenseId)
  );
  return (
    <EditExpensePage
      {...props}
      navigate={navigate}
      expense={expense}
    />
  );
};

EditExpensePage.propTypes = {
  expense: PropTypes.object.isRequired,
  navigate: PropTypes.func,
  startEditExpense: PropTypes.func,
  startRemoveExpense: PropTypes.func,
};

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (expenseId, expenseUpdate) => {
    dispatch(startEditExpense(expenseId, expenseUpdate));
  },
  startRemoveExpense: ({ id }) => {
    dispatch(startRemoveExpense({ id }));
  },
});

export default connect(undefined, mapDispatchToProps)(EditExpensePageWrapper);
