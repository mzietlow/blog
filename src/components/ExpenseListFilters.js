import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };

  onFocusChange = (calendarFocused) =>
    this.setState(() => ({ calendarFocused }));

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => this.props.dispatch(setTextFilter(e.target.value))}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={({ target }) =>
            this.props.dispatch(chooseSortBy(target.value))
          }
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        ></DateRangePicker>
      </div>
    );
  }
}

const chooseSortBy = (selectVal) => {
  switch (selectVal) {
    case "date":
      return sortByDate();
    case "amount":
      return sortByAmount();
  }
};

ExpenseListFilters.propTypes = {
  filters: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
};

// Todo: PropTypes setzen @ 4:20 https://www.udemy.com/course/react-2nd-edition/learn/lecture/7900120#overview
const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(mapStateToProps)(ExpenseListFilters);
