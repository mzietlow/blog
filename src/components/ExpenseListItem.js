import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
const ExpenseListItem = ({
  id,
  description,
  note,
  amount,
  createdAt,
}) => {
  return (
    <div>
    <Link to={`/edit/${id}`}><h2>{description}</h2></Link>
      <p>note: {note}</p>
      <p>amount: {amount}</p>
      <p>createdAt: {createdAt}</p>
    </div>
  );
};

ExpenseListItem.propTypes = {
  expenses: PropTypes.object,
  description: PropTypes.string,
  note: PropTypes.string,
  amount: PropTypes.number,
  createdAt: PropTypes.number,
  dispatch: PropTypes.func,
  id: PropTypes.string,
};

export default ExpenseListItem;
