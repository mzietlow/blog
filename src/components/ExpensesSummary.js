import React from "react";
import PropTypes from "prop-types";
import getTotalExpense from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import { connect } from "react-redux";
import numeral from "numeral";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount == 1 ? "expense" : "expenses";
  const formattedExpensesTotal = numeral(expensesTotal / 100).format("$0,0.00");
  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling{" "}
        {formattedExpensesTotal}
      </h1>
    </div>
  );
};

ExpensesSummary.propTypes = {
  expenseCount: PropTypes.number,
  expensesTotal: PropTypes.number,
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses({ ...state });
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: getTotalExpense(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
