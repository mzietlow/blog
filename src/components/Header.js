import React from 'react'
import PropTypes from 'prop-types'

const Header = (props) => {
  return (
    <div>
      <h1>{props.title}</h1>
      <h2>{props.subtitle}</h2>
    </div>
  );
};

Header.defaultProps = {
  title: "someDefault",
};

Header.propTypes = {
  title: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
};

export default Header;