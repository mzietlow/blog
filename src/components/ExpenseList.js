/* eslint-disable react/prop-types */
import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses'

const ExpenseList = (props) => (
  <div>
    <h1>Expense List</h1>
    {props.expenses.map((expense) => {
      return <ExpenseListItem key={expense.id} {...expense} />;
    })}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses({...state}),
  };
};

export default connect(mapStateToProps)(ExpenseList);
