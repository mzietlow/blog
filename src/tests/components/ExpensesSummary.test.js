import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";


test("should render ExpensesSummary with 0 expenses", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={0} expensesTotal={0}/>);
  expect(wrapper).toMatchSnapshot();
});
test("should render ExpensesSummary with 1 expense", () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={123} />);
  expect(wrapper).toMatchSnapshot();
});
test("should render ExpensesSummary with 2 expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={2} expensesTotal={12.3} />
  );
  expect(wrapper).toMatchSnapshot();
});
