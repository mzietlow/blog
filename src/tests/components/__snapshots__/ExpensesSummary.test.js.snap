// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 0 expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     totalling
     
    $0.00
  </h1>
</div>
`;

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling
     
    $1.23
  </h1>
</div>
`;

exports[`should render ExpensesSummary with 2 expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     totalling
     
    $0.12
  </h1>
</div>
`;
